contributors/devel/code-generation-outside-kubernetes.md

```diff
@@ -0,0 +1,54 @@
+# Code generation (from the experience of developing service-catalog)
+
+## Code generator binaries
+There are several libraries in the Kubernetes tree that are responsible for
+various code generation operations. Those go based libraries are currently
+located at cmd/libs/go2idl/. In each directory, there is a main.go file
+that currently has hard coded an internal path for boilerplate.go.txt.
+However, there now exists a repo-infra repository that can be used instead.
+These files serve mostly as wrappers for executing and utilizing gengo.
+
+## Code generation arguments
+In Kubernetes, most of the code generation execution handling is in the
+Makefile.generated_files and is in the .generated_files target. The rule
+complexity is due to Kubernetes generating a DAG in order to determine
+which files need rebuilt and code regenerated for.
+
+Each code generation binary supports different arguments. The arguments of
+importance are:
+--extra-peer-dirs - defines additional locations to provide current operations
+--bounding-dirs - list of import paths for binding the types
+--input-dirs - list of import paths to get input types from
+
+In service-catalog, the makefile has been significantly simplified and one
+can see the parameters sent to each code generator here:
+
+https://github.com/kubernetes-incubator/service-catalog/blob/59cd78081f6cb23aa102df493f5b2471e8e0cdd2/Makefile#L102
```

Comment:
I don't understand what you mean by three different projects, but the makefile section that I was linking to above is practically the same today. I tend to like the linking here, but can pull the makefile content in if that would help understanding.